src/framebuffer.rs

```diff
@@ -0,0 +1,707 @@
+            RawU4, RawU8, ToBytes,
+        },
+        PixelColor,
+    },
+    Pixel,
+};
+
+/// Calculates the required buffer size.
+///
+/// This function is a workaround for current limitations in Rust const generics.
+/// It can be used to calculate the `N` parameter based on the size and color type of the framebuffer.
+pub const fn buffer_size<C: PixelColor>(width: usize, height: usize) -> usize {
+    (width * C::Raw::BITS_PER_PIXEL + 7) / 8 * height
+}
+
+/// A framebuffer.
+///
+/// # Examples
+///
+/// ```
+/// use embedded_graphics::{
+///     framebuffer,
+///     framebuffer::{Framebuffer, buffer_size},
+///     pixelcolor::{Rgb565, raw::LittleEndian},
+///     prelude::*,
+///     primitives::PrimitiveStyle,
+/// };
+///
+/// let mut fb = <framebuffer!(Rgb565, 320, 240)>::new();
+/// // or:
+/// let mut fb = Framebuffer::<Rgb565, _, LittleEndian, 320, 240, {buffer_size::<Rgb565>(320, 240)}>::new();
+///
+/// fb.bounding_box()
+///     .into_styled(PrimitiveStyle::with_stroke(Rgb565::RED, 1))
+///     .draw(&mut fb)
+///     .unwrap();
+/// ```
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub struct Framebuffer<C, R, BO, const WIDTH: usize, const HEIGHT: usize, const N: usize> {
+    data: [u8; N],
+    color_type: PhantomData<C>,
+    raw_type: PhantomData<R>,
+    byte_order: PhantomData<BO>,
+}
+
+impl<C, BO, const WIDTH: usize, const HEIGHT: usize, const N: usize>
+    Framebuffer<C, C::Raw, BO, WIDTH, HEIGHT, N>
+where
+    C: PixelColor,
+{
+    /// Static assertion that N is correct.
+    /// MSRV: remove N when constant generic expressions are stabilized
+    const CHECK_N: () = if N != buffer_size::<C>(WIDTH, HEIGHT) {
+        panic!("Invalid N: see Framebuffer documentation for more information");
+    };
+
+    /// Creates a new framebuffer.
+    pub fn new() -> Self {
+        #[allow(path_statements)]
+        {
+            // Make sure CHECK_N isn't optimized out.
+            Self::CHECK_N;
+        }
+
+        Self {
+            data: [0; N],
+            color_type: PhantomData,
+            raw_type: PhantomData,
+            byte_order: PhantomData,
+        }
+    }
+
+    /// Returns a reference to the raw framebuffer data.
+    pub fn data(&self) -> &[u8; N] {
+        &self.data
+    }
+
+    /// Returns a mutable reference to the raw framebuffer data.
+    pub fn data_mut(&mut self) -> &mut [u8; N] {
+        &mut self.data
+    }
+}
+
+impl<C, BO, const WIDTH: usize, const HEIGHT: usize, const N: usize>
+    Framebuffer<C, C::Raw, BO, WIDTH, HEIGHT, N>
+where
+    C: PixelColor + From<C::Raw>,
+    BO: ByteOrder,
+    for<'a> RawDataSlice<'a, C::Raw, BO>: IntoIterator<Item = C::Raw>,
+{
+    /// Returns an image based on the framebuffer content.
+    pub fn as_image(&self) -> ImageRaw<'_, C, BO> {
+        ImageRaw::new(&self.data, WIDTH as u32)
+    }
+
+    /// Get pixel color.
+    ///
+    /// Returns `None` is `p` is outside the framebuffer bounding box.
```

Comment:
Mostly a note to self for a future PR unless you beat me to it. I think we could get some pretty good performance gains here.

```suggestion
    /// Returns `None` is `p` is outside the framebuffer bounding box.
    // TODO: Optimise implementation by moving away from `ImageRaw::pixel` and directly calculating skip value.
```

Reply:
Agreed. I wanted to add a similar comment, but forgot to add it in the end. But this is also something that should be benchmarked before changing it IMO.